### PR TITLE
Inference autoscaling telemetry

### PR DESCRIPTION
**Pull request by jan-elastic**
This PR add telemetry for inference adaptive autoscaling.

This feature needs gauge metrics labeled by deployment_id, comparable to the examples listed here:
https://opentelemetry.io/docs/specs/otel/metrics/api/#asynchronous-gauge-creation

Unfortunately, the existing telemetry framework in ES for gauges only allows single values, therefore I had to extend it.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java [hunk 1]
```diff
@@ -75,6 +82,108 @@ Stats sub(Stats value) {
...
+
+        private final List<AutoCloseable> metrics = new ArrayList<>();
+
+        Metrics() {}
+
+        void init() {
+            if (metrics.isEmpty() == false) {
+                return;
+            }
+            metrics.add(
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java:95]: Please add a metric for the deployment total inference count which is also in the stats API response.

Comment (jan-elastic) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java:95]: As discussed offline: this is very useful, but not as part of adaptive allocations. Let's implement it somewhere else with different metrics name, etc.

**Review by pgomulka — COMMENTED** [review]
I think this is something a broader team would like to discuss hence pinging @elastic/es-core-infra 
My main worry would be if we want to change the signature of the methods to accept the collection of attributes or have a variant of methods

**Comment by jan-elastic** [comment]
> I think this is something a broader team would like to discuss hence pinging @elastic/es-core-infra My main worry would be if we want to change the signature of the methods to accept the collection of attributes or have a variant of methods

Thanks for bringing them in the loop.

Note that the signatures of the public-facing methods (like `registerDoubleGauge`) have not changed, and I've added variants (like `registerDoublesGauge` (plural)).

Regarding the internal methods (like `DoubleAsyncCounterAdapter`), I think the current approach leads to less code duplication without disadvantages, but I'll leave that decision up to the @elastic/es-core-infra team.

**Comment by jan-elastic** [comment]
@pgomulka @elastic/es-core-infra Could you have a look at this PR? If possible, I'd like to wrap this up this week before I go on holiday. Thanks!

**Review by mosche — APPROVED** [review]
Sorry, for the delay on this @jan-elastic!
We've just discussed your addition to the metrics interface in core/infra, that part looks good to us 👍 Thanks for the contribution :)

**Comment by jan-elastic** [comment]
Thanks for the review!

**Pull request merged**